src/InfluxDB.Collector/Util/PseudoHighResTimestampSource.cs

```diff
@@ -0,0 +1,61 @@
+﻿using System;
+
+namespace InfluxDB.Collector.Util
+{
+
+
+
+    /// <summary>
+    /// Implements <see cref="ITimestampSource"/>
+    /// in a way that combines the low-ish resolution DateTime.UtcNow
+    /// with a sequence number added to the ticks to provide 
+    /// pseudo-tick precision timestamps
+    /// </summary>
+    /// <remarks>
+    /// See https://github.com/influxdata/influxdb-csharp/issues/46 for why this is necessary.
+    /// Long story short: 
+    /// a) InfluxDB has a "LastWriteWins" policy for points that have the same timestamp and tags. 
+    /// b) The normal <see cref="System.DateTime.UtcNow"/> only supplies timestamps that change not as often as you think.
+    /// c) In a web server, it's entirely possible for more than one thread to get the same UtcNow value
+    /// 
+    /// As a remediation for this, we infuse DateTime.UtcNow with a sequence number until it ticks over.
+    /// 
+    /// </remarks>
+    public class PseudoHighResTimestampSource : ITimestampSource
+    {
+
+        private long _lastUtcNowTicks = 0;
+        private long _sequence = 0;
+        private readonly object lockObj = new object();
+
+       
+        public DateTime GetUtcNow()
+        {
+            DateTime utcNow = DateTime.UtcNow;
+
+            lock (lockObj)
```

Comment:
Locking on every metric is fine if we have to do it, but it seems like we could potentially avoid it using some interlocked methods?

Reply:
I initially started out with `Interlocked` but my late evening brain couldn't devise a way to make the logic work (i.e. we need to do three atomic thread safe operations - check the last value, branch (a) if the same (increment), branch (b) if not the same (reset)). 

There's _maybe_ an operation here that sees a double check (`if...lock...if` pattern) but it'd need to be benchmarked to see if it's actually a saving over locking. Would love to see an alternative.

Reply:
Its possible using a solution like the first answer here: https://stackoverflow.com/questions/5608980/how-to-ensure-a-timestamp-is-always-unique

I've implemented a similar solution using that approach successfully, but I personally haven't benchmarked it against a lock based solution like yours.

Reply:
Happy to go with the working lock-based fix and explore the lock-free option afterwards, if it would help unblock things.

Locking is tricky to benchmark since perf depends heavily on contention and other non-local effects on the app. I personally run with the rule of thumb that locking is better to avoid in frequently-called library code when possible - may not always be the case, though :-)

Reply:
@optical I did a quick benchmark locally with both mechanisms and while ymmv, I found the interlocked method slower than the locked method _for parallel execution in tight loops_, but marginally faster for _serial loops with low contention_.

100,000,000 iterations for 2 implementations of GetUtcNow() - Interlocked and Lock. Parallel are on 4 cores.

 Method            | Serial        | Parallel 
------------------|------------|----------
Loop overhead  | 335ms      | 211ms
Locking              | 4812ms    | 5327ms
Interlocked        | 4155ms    | 13653ms

The reason why the interlocked method looks so very much worse in the parallel case is the high amount of contention in a tight loop causes the inner loop (used to generate the sequence number) to have to run a long time (akin to a Shlemiel's painter algorithm). If the contention is a lot lower (as in the serial case, or probably reality :-)) then the performance is back to being comparable. 

I'd take the more predictable performance of the locking method than the unpredictable loop+interlock method. It is splitting hairs though. Both methods run fast enough not to worry about IMHO.